specs/NotifyDropTargetAction.md

```diff
@@ -0,0 +1,313 @@
+{
+  webView2CompositionController.DragLeave();
+}
+
+private void WebView_Drop(object sender, DragEventArgs e)
+{
+  uint keyboardState =
+    ConvertDragDropModifiersToWin32KeyboardState(e.Modifiers);
+  Point pointerPosition = CoreWindow.GetForCurrentThread().PointerPosition;
+  DataPackageOperation operation =
+    (DataPackageOperation)webView2CompositionController.Drop(
+      e.Data, keyboardState, pointerPosition);
+  e.AcceptedOperation = operation;
+}
+
+// Win32 keyboard state modifiers that are relevant during drag and drop.
+public const uint MK_LBUTTON = 1;
+public const uint MK_RBUTTON = 2;
+public const uint MK_SHIFT = 4;
+public const uint MK_CONTROL = 8;
+public const uint MK_MBUTTON = 16;
+public const uint MK_ALT = 32;
+
+// Helper function to convert DragDropModifiers to win32 keyboard state
+// modifiers that WebView2 uses during drag and drop operation.
+private uint ConvertDragDropModifiersToWin32KeyboardState(
+  Windows.ApplicationModel.DataTransfer.DragDrop.DragDropModifiers modifiers)
+{
+  uint win32DragDropModifiers = 0;
+  if ((modifiers & DragDropModifiers.Shift) == DragDropModifiers.Shift)
+    win32DragDropModifiers |= MK_SHIFT;
+  if ((modifiers & DragDropModifiers.Control) == DragDropModifiers.Control)
+    win32DragDropModifiers |= MK_CONTROL;
+  if ((modifiers & DragDropModifiers.Alt) == DragDropModifiers.Alt)
+    win32DragDropModifiers |= MK_ALT;
+  if ((modifiers & DragDropModifiers.LeftButton) == DragDropModifiers.LeftButton)
+    win32DragDropModifiers |= MK_LBUTTON;
+  if ((modifiers & DragDropModifiers.MiddleButton) == DragDropModifiers.MiddleButton)
+    win32DragDropModifiers |= MK_MBUTTON;
+  if ((modifiers & DragDropModifiers.RightButton) == DragDropModifiers.RightButton)
+    win32DragDropModifiers |= MK_RBUTTON;
+  return win32DragDropModifiers;
+}
+```
+
+
+# API Details
+## Win32
+```c++
+interface ICoreWebView2CompositionController2 : ICoreWebView2CompositionController {
+  /// This set of APIs (DragEnter, DragLeave, DragOver, and Drop) will allow
+  /// users to drop things such as images, text, and links into the WebView as
+  /// part of a drag/drop operation. The reason that we need a separate API for
+  /// this with composition hosting is because we don't have an HWND to call
+  /// RegisterDragDrop on. The hosting app needs to call RegisterDragDrop on the
+  /// HWND that contains the WebView and implement IDropTarget so it can forward
+  /// the calls (IDropTarget::DragEnter, DragMove, DragLeave, and Drop) to the
+  /// WebView.
+  ///
+  /// This function corresponds to IDropTarget::DragEnter
+  ///
+  /// The hosting application must register as an IDropTarget and implement
+  /// and forward DragEnter calls to this function.
+  ///
+  /// In addition, the hosting application needs to create an IDropTargetHelper
+  /// and call the corresponding IDropTargetHelper::DragEnter function on that
+  /// object before forwarding the call to WebView.
```

Comment:
Why is the IDropTargetHelper required? I mean, I know why it's good, but what goes wrong if you don't do it?

Reply:
I originally added it because I thought it would be needed to update the cursor. However, after removing it from our sample code, all of the images over the cursor still seems to be updating correctly. It seems like it to me it's the source has more of the responsibility to update the cursor image. Now I'm unsure of what IDropTargetHelper does.

Reply:
Debugged through the Chromium code and it seems like it does call the IDropTargetHelper and it does work even in another process. I don't think we need it in the sample anymore.